lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,683 @@
+        if (!internalPath) {
+          blackBox.extModule.emitOpError(
+              llvm::formatv("DataTapModuleSignalKey annotation on port {0} "
+                            "missing \"internalPath\" attribute",
+                            portName));
+          signalPassFailure();
+          continue;
+        }
+
+        wiring.prefices = instanceGraph.getAbsolutePaths(op);
+        wiring.suffix = internalPath.getValue();
+        portWiring.push_back(std::move(wiring));
+        continue;
+      }
+
+      // Handle data taps with literals.
+      if (cls == strings.literalKeyClass) {
+        blackBox.extModule.emitOpError(llvm::formatv(
+            "LiteralDataTapKey annotations not yet supported (on port {0})",
+            portName));
+        signalPassFailure();
+        continue;
+      }
+
+      // Handle memory taps.
+      if (cls == strings.memTapClass) {
+        auto op = tappedOps.lookup(portAnno.anno);
+        if (!op) {
+          signalPassFailure();
+          llvm::report_fatal_error(
+              llvm::formatv(
+                  "DataTapModuleSignalKey annotation was not scattered to "
+                  "an operation: {0}",
+                  portAnno.anno)
+                  .str());
+        }
+
+        // Extract the name of the memory.
+        // TODO: This would be better handled through a proper cross-module
+        // reference preserved in the IR, such that ExportVerilog can insert a
+        // proper name here at the last moment.
+        auto name = op->getAttrOfType<StringAttr>("name");
+        if (!name) {
+          auto diag = op->emitError("target of memory tap must have a name");
+          diag.attachNote(blackBox.extModule->getLoc()) << llvm::formatv(
+              "used by MemTapAnnotation on port {0} here:", portName);
+          signalPassFailure();
+          continue;
+        }
+
+        // Formulate a hierarchical reference into the memory.
+        // CAVEAT: This just assumes that the memory will map to something that
+        // can be indexed in the final Verilog. If the memory gets turned into
+        // an instance of some sort, we lack the information necessary to go in
+        // and access individual elements of it. This will break horribly since
+        // we generate memory impls out-of-line already, and memories coming
+        // from an external generator are even worse. This needs a special node
+        // in the IR that can properly inject the memory array on emission.
+        wiring.prefices =
+            instanceGraph.getAbsolutePaths(op->getParentOfType<FModuleOp>());
+        wiring.suffix = llvm::formatv("{0}[{1}]", name.getValue(),
+                                      memPortIdx[portAnno.anno]++);
+        portWiring.push_back(std::move(wiring));
+        continue;
+      }
+
+      llvm_unreachable(
+          llvm::formatv("unknown annotation {}", portAnno.anno).str().c_str());
```

Comment:
Passes shuldn't crash on malformed input.  Please emitOpError

Reply:
This is truly unreachable, because the code above that populates the list of annos being processed here only adds the different classes handled here. I've changed the `llvm_unreachable` to make this clear (and get rid of the `formatv`).